File: android-beans/src/main/java/rocks/frieler/android/beans/BeanRegistry.java
package rocks.frieler.android.beans;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.TreeMap;

import rocks.frieler.android.beans.scopes.ScopedFactoryBean;
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanDecorator;
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler;
import rocks.frieler.android.beans.scopes.prototype.PrototypeScopedFactoryBeanHandler;
import rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBeanHandler;


/**
 * {@link BeansProvider} that holds and provides beans which were explicitly registered before.
 * <p>
 * The {@link BeanRegistry} supports the {@value SingletonScopedFactoryBeanHandler#SINGLETON_SCOPE} and
 * {@value PrototypeScopedFactoryBeanHandler#PROTOTYPE_SCOPE} by default.
 */
public class BeanRegistry implements BeansProvider {
    private final Map<String, Object> beans = new HashMap<>();
    private final List<BeanPostProcessor> beanPostProcessors = new LinkedList<>();

    private final Map<String, ScopedFactoryBeanHandler> beanScopes = new TreeMap<>();

    BeanRegistry() {
        addBeanScope(new SingletonScopedFactoryBeanHandler());
        addBeanScope(new PrototypeScopedFactoryBeanHandler());
    }

    /**
     * Adds a bean-scope expressed by the given {@link ScopedFactoryBeanHandler} to this {@link BeanRegistry}.
     *
     * @param scopedFactoryBeanHandler the {@link ScopedFactoryBeanHandler} to add
     */
    void addBeanScope(ScopedFactoryBeanHandler scopedFactoryBeanHandler) {
        beanScopes.put(scopedFactoryBeanHandler.getName(), scopedFactoryBeanHandler);
    }

    /**
     * Registers the bean with a generated name at this {@link BeanRegistry}.
     *
     * @param bean the bean
     */
    void registerBean(Object bean) {
        registerBean(generateBeanName(bean), bean);
    }

    private String generateBeanName(Object bean) {
        Class<?> beanClass;
        if (bean instanceof ScopedFactoryBean) {
            beanClass = ((ScopedFactoryBean) bean).getBeanType();
        } else {
            beanClass = bean.getClass();
        }

        String preferredBeanName = getPreferredBeanName(beanClass);
        if (!beans.containsKey(preferredBeanName)) {
            return preferredBeanName;
        }

        int i = 2;
        while (beans.containsKey(preferredBeanName + i)) {
            i++;
        }
        return preferredBeanName + i;
    }

    /**
     * Registers the bean with the given name at this {@link BeanRegistry}.
     * <p>
     * The bean will be post-processed by all registered {@link BeanPostProcessor}s.
     * <p>
     * If a bean with this name already exists, it will be replaced.
     *
     * @param name the name of the bean
     * @param bean the bean
     */
    void registerBean(String name, Object bean) {
        bean = postProcessBean(name, bean);
        beans.put(name, bean);
    }

    /**
     * Registers the given {@link BeanPostProcessor}.
     * <p>
     * The {@link BeanPostProcessor} will get all beans already registered to post-process immediately and all beans
     * registered subsequently.
     *
     * @param beanPostProcessor the new {@link BeanPostProcessor}
     */
    void registerBeanPostProcessor(BeanPostProcessor beanPostProcessor) {
        this.beanPostProcessors.add(beanPostProcessor);

        for (Entry<String, Object> beanEntry : beans.entrySet()) {
            beanEntry.setValue(beanPostProcessor.postProcessBean(beanEntry.getKey(), beanEntry.getValue()));
        }
    }

    private <T> T postProcessBean(String name, T bean) {
        for (BeanPostProcessor postProcessor : beanPostProcessors) {
            bean = postProcessor.postProcessBean(name, bean);
        }
        return bean;
    }

    @Override
    public <T> T lookUpBean(String name, Class<T> type) {
        Object beanCandidate = beans.get(name);
        if (beanCandidate != null) {
            T bean = resolveBeanFromCandidate(name, type, beanCandidate);
            if (bean != null) {
                return bean;
            }
        }
        return null;
    }

    @Override
    public <T> T lookUpBean(Class<T> type) {
        String preferredBeanName = getPreferredBeanName(type);
        Object beanCandidateByPreferredName = beans.get(preferredBeanName);
        if (beanCandidateByPreferredName != null) {
            T bean = resolveBeanFromCandidate(preferredBeanName, type, beanCandidateByPreferredName);
            if (bean != null) {
                return bean;
            }
        }

        for (Entry<String, Object> beanCandidate : beans.entrySet()) {
            T bean = resolveBeanFromCandidate(beanCandidate.getKey(), type, beanCandidate.getValue());
            if (bean != null) {
                return bean;
            }
        }
        return null;
    }

    @Override
    public <T> List<T> lookUpBeans(Class<T> type) {
        List<T> matchingBeans = new ArrayList<>();
        for (Entry<String, Object> beanCandidate : beans.entrySet()) {
            T bean = resolveBeanFromCandidate(beanCandidate.getKey(), type, beanCandidate.getValue());
            if (bean != null) {
                matchingBeans.add(bean);
            }
        }
        return matchingBeans;
    }

    @Nullable
    private <T> T resolveBeanFromCandidate(String name, Class<T> type, Object beanCandidate) {
        if (beanCandidate instanceof ScopedFactoryBean && type.isAssignableFrom(((ScopedFactoryBean<?>)beanCandidate).getBeanType())) {
            @SuppressWarnings("unchecked") final ScopedFactoryBean<T> factoryBean = (ScopedFactoryBean<T>) beanCandidate;
            final ScopedFactoryBeanHandler scopedFactoryBeanHandler = beanScopes.get(factoryBean.getScope());
            if (scopedFactoryBeanHandler != null && scopedFactoryBeanHandler.isActive()) {
                return scopedFactoryBeanHandler.getBean(name, ScopedFactoryBeanDecorator.Companion.decorate(factoryBean).withPostProcessing((bean) -> postProcessBean(name, bean)));
            }
        }

        if (type.isAssignableFrom(beanCandidate.getClass())) {
            //noinspection unchecked
            return (T) beanCandidate;
        }

        return null;
    }

    public static String getPreferredBeanName(Class<?> beanClass) {
        return beanClass.getName();
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/BeanRegistry.kt
package rocks.frieler.android.beans

import rocks.frieler.android.beans.scopes.ScopedFactoryBean
import rocks.frieler.android.beans.scopes.ScopedFactoryBeanHandler
import rocks.frieler.android.beans.scopes.decorate
import rocks.frieler.android.beans.scopes.prototype.PrototypeScopedFactoryBeanHandler
import rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBeanHandler
import rocks.frieler.kotlin.reflect.isAssignableFrom
import java.util.LinkedList
import java.util.TreeMap
import kotlin.reflect.KClass
import kotlin.reflect.jvm.jvmName

/**
 * [BeansProvider] that holds and provides beans which were explicitly registered before.
 *
 * The [BeanRegistry] supports the
 * [singleton][SingletonScopedFactoryBeanHandler.SINGLETON_SCOPE] and
 * [prototype][PrototypeScopedFactoryBeanHandler.PROTOTYPE_SCOPE] scopes by default.
 *
 * [BeanRegistry] is also a [HierarchicalBeansProvider], so it supports a parent to inherit beans
 * from.
 *
 * @author Christopher Frieler
 */
class BeanRegistry internal constructor(parent: BeansProvider? = null) : HierarchicalBeansProvider(parent) {
    private val beans: MutableMap<String, Any> = HashMap()
    private val beanPostProcessors: MutableList<BeanPostProcessor> = LinkedList()

    private val beanScopes: MutableMap<String, ScopedFactoryBeanHandler> = TreeMap()

    init {
        addBeanScope(SingletonScopedFactoryBeanHandler())
        addBeanScope(PrototypeScopedFactoryBeanHandler())
    }

    /**
     * Adds a bean-scope expressed by the given [ScopedFactoryBeanHandler] to this [BeanRegistry].
     *
     * @param scopedFactoryBeanHandler the [ScopedFactoryBeanHandler] to add
     */
    fun addBeanScope(scopedFactoryBeanHandler: ScopedFactoryBeanHandler) {
        beanScopes[scopedFactoryBeanHandler.name] = scopedFactoryBeanHandler
    }

    /**
     * Registers the bean with a generated name at this [BeanRegistry] and returns this name.
     *
     * @param bean the bean
	 * @return the generated name for the bean
     */
    fun registerBean(bean: Any): String {
		val name = generateBeanName(bean)
		registerBean(name, bean)
		return name
    }

    private fun generateBeanName(bean: Any): String {
        val beanClass: KClass<*> = if (bean is ScopedFactoryBean<*>) bean.beanType else bean::class

        var beanName = getPreferredBeanName(beanClass)
        if (beans.containsKey(beanName)) {
			var suffix = 2
			while (beans.containsKey(beanName + suffix)) {
				suffix++
			}
			beanName += suffix
        }

        return beanName
    }

    /**
     * Registers the bean with the given name at this [BeanRegistry].
     *
     *
     * The bean will be post-processed by all registered [BeanPostProcessor]s. If the bean is a
	 * [BeanPostProcessor] itself, it will receive all beans already registered to post-process
	 * immediately and all beans registered subsequently, but never the bean itself.
     *
     *
     * If a bean with this name already exists, it will be replaced.
     *
     * @param name the name of the bean
     * @param bean the bean
     */
    fun registerBean(name: String, bean: Any) {
		val effectiveBean = postProcessBean(name, bean)

		if (effectiveBean is BeanPostProcessor) {
			registerBeanPostProcessor(effectiveBean)
		}

		beans[name] = effectiveBean
    }

    private fun registerBeanPostProcessor(beanPostProcessor: BeanPostProcessor) {
        beanPostProcessors.add(beanPostProcessor)

        for (beanEntry in beans.entries) {
            beanEntry.setValue(beanPostProcessor.postProcessBean(beanEntry.key, beanEntry.value))
        }
    }

    private fun <T :Any> postProcessBean(name: String, bean: T): T {
        var postProcessedBean = bean
        for (postProcessor in beanPostProcessors) {
            postProcessedBean = postProcessor.postProcessBean(name, postProcessedBean)
        }
        return postProcessedBean
    }

	override fun <T : Any> lookUpOptionalLocalBean(name: String, type: KClass<T>): T? {
        val beanCandidate = beans[name]
        return if (beanCandidate == null) {
            null
        } else {
            resolveBeanFromCandidate(name, type, beanCandidate)
        }
	}

    override fun <T :Any> lookUpOptionalLocalBean(type: KClass<T>): T? {
        val preferredBeanName = getPreferredBeanName(type)
        val beanCandidateByPreferredName = beans[preferredBeanName]
        if (beanCandidateByPreferredName != null) {
            val bean = resolveBeanFromCandidate(preferredBeanName, type, beanCandidateByPreferredName)
            if (bean != null) {
                return bean
            }
        }

        for ((key, value) in beans) {
            val bean = resolveBeanFromCandidate(key, type, value)
            if (bean != null) {
                return bean
            }
        }

        return null
    }

    override fun <T :Any> lookUpLocalBeans(type: KClass<T>): List<T> {
        val matchingBeans: MutableList<T> = ArrayList()
        for ((key, value) in beans) {
            val bean = resolveBeanFromCandidate(key, type, value)
            if (bean != null) {
                matchingBeans.add(bean)
            }
        }
        return matchingBeans
    }

	private fun <T :Any> resolveBeanFromCandidate(name: String, type: KClass<T>, beanCandidate: Any): T? {
		if (beanCandidate is ScopedFactoryBean<*> && type.isAssignableFrom(beanCandidate.beanType)) {
			@Suppress("UNCHECKED_CAST") val factoryBean = beanCandidate as ScopedFactoryBean<T>
            val scopedFactoryBeanHandler = beanScopes[factoryBean.scope]
            if (scopedFactoryBeanHandler != null && scopedFactoryBeanHandler.isActive) {
                return scopedFactoryBeanHandler.getBean(name, factoryBean.decorate().withPostProcessing { bean: T -> postProcessBean(name, bean) }, this)
            }
        }

		@Suppress("UNCHECKED_CAST")
		return if (type.isInstance(beanCandidate)) beanCandidate as T else null
	}

    private companion object BeanNaming {
		fun getPreferredBeanName(beanClass: KClass<*>): String {
			return beanClass.jvmName
        }
    }
}
